service/triereader/parser.go

```diff
@@ -12,51 +12,46 @@
 // License for the specific language governing permissions and limitations under
 // the License.
 
-package feeder
+package triereader
 
 import (
 	"fmt"
 
 	"github.com/onflow/flow-go/ledger"
 	"github.com/onflow/flow-go/ledger/complete/wal"
-
-	"github.com/onflow/flow-archive/models/archive"
 )
 
-// Feeder is a component that retrieves trie updates and feeds them to its consumer.
-type Feeder struct {
+// WalParser is a component that retrieves trie updates and feeds them to its consumer.
+type WalParser struct {
 	reader WALReader
 }
 
-// FromWAL creates a trie update feeder that sources state deltas from a WAL reader.
-func FromWAL(reader WALReader) *Feeder {
+// FromWAL creates a trie update triereader that sources state deltas from a WAL reader.
+func FromWAL(reader WALReader) *WalParser {
 
-	f := Feeder{
+	f := WalParser{
 		reader: reader,
 	}
 
 	return &f
 }
 
-// Update returns the next trie update.
-func (f *Feeder) Update() (*ledger.TrieUpdate, error) {
+// AllUpdates returns all updates for the entire spork in wal.
+func (f *WalParser) AllUpdates() ([]*ledger.TrieUpdate, error) {
 
 	// We read in a loop because the WAL contains entries that are not trie
 	// updates; we don't really need to care about them, so we can just skip
 	// them until we find a trie update.
-	for {
+	updates := make([]*ledger.TrieUpdate, 0)
+	for f.reader.Next() {
 
 		// This part reads the next entry from the WAL, makes sure we didn't
 		// encounter an error when reading or decoding and ensures that it's a
 		// trie update.
-		next := f.reader.Next()
 		err := f.reader.Err()
```

Comment:
This felt weird to me.
Why would you check for error inside the `.Next` loop. 

So I went and checked, and if you look at 
wal/reader.go:56

```
// Next advances the reader to the next records and returns true if it exists.
// It must not be called again after it returned false.
func (r *Reader) Next() bool {
	err := r.next()
	if errors.Cause(err) == io.EOF {
		// The last WAL segment record shouldn't be torn(should be full or last).
		// The last record would be torn after a crash just before
		// the last record part could be persisted to disk.
		if r.curRecTyp == recFirst || r.curRecTyp == recMiddle {
			r.err = errors.New("last record is torn")
		}
		return false
	}
	r.err = err
	return r.err == nil   //   <----------
}
```

next will be false if there is an error. So error handling should be outside of the loop.

I think the `WALReader` interface should also note that in the case of error in `Next`, `Next` should return false.